src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -1109,19 +1110,27 @@ void AsyncRPCOperation_sendmany::add_taddr_outputs_to_tx() {
     tx_ = CTransaction(rawTx);
 }
 
-void AsyncRPCOperation_sendmany::add_taddr_change_output_to_tx(CAmount amount) {
+void AsyncRPCOperation_sendmany::add_taddr_change_output_to_tx(CAmount amount, bool change) {
 
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
     EnsureWalletIsUnlocked();
-    CReserveKey keyChange(pwalletMain);
-    CPubKey vchPubKey;
-    bool ret = keyChange.GetReservedKey(vchPubKey);
-    if (!ret) {
-        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Could not generate a taddr to use as a change address"); // should never fail, as we just unlocked
+    CTxOut out;
+    if (change) {
+        CScript scriptPubKey = GetScriptForDestination(fromtaddr_.Get());
```

Comment:
Do we have considered cases where fromtaddr does not represent a pubkey? It could be useful to check it and not proceeding if not. See https://github.com/ZencashOfficial/zen/blob/master/src/base58.cpp#L286

Reply:
Fromtaddr can be either [IsPubKey](https://github.com/ZencashOfficial/zen/blob/314b7701d945d71f4ee362a0ad99a6e52f2d432c/src/base58.cpp#L296) or [IsScript](https://github.com/ZencashOfficial/zen/blob/314b7701d945d71f4ee362a0ad99a6e52f2d432c/src/base58.cpp#L303), `GetScriptForDestination` returns the correct scriptPubKey for both. P2SH addresses have been considered and a testcase has been added for them, but additional validation wouldn't hurt.